### PR TITLE
Use PPX-Yojson for the configuration

### PR DESCRIPTION
**Pull request by smondet**
This PR goes on top of https://github.com/hammerlab/ketrew/pull/142.

I've uploaded the documentation corresponding to this branch:
http://seb.mondet.org/software/ketrew/json_config/index.html

Fixes https://github.com/hammerlab/ketrew/issues/124 and https://github.com/hammerlab/ketrew/issues/123.

[Review on Reviewable](https://reviewable.io/reviews/hammerlab/ketrew/143)

### CODE REVIEW

File: src/doc/The_Configuration_File.md [hunk 1]
```diff
@@ -1,147 +1,184 @@
 Configuration File
 ==================
```
Comment (rleonid) [inline src/doc/The_Configuration_File.md]: I put together my proposed changes to this file in https://github.com/hammerlab/ketrew/tree/json_config_the_doc. Some are stylistic/wording choices so not necessary to merge the entire thing, do as you see fit.

File: src/lib/ketrew_command_line.ml [hunk 2]
```diff
@@ -559,17 +560,16 @@ let cmdliner_main ?override_configuration ?argv ?(additional_commands=[]) () =
...
+        pure (fun configuration ->
             match Configuration.mode configuration with
-            | `Server srv -> Ketrew_server.start srv
+            | `Server srv ->
+              daemonize_if_applicable
+                Ketrew_configuration.(
+                  match daemon srv, log_path srv  with
+                  | true, popt -> `Daemonize_with popt
+                  | false, _ -> `Do_not_daemonize
+                );
```
Comment (rleonid) [inline src/lib/ketrew_command_line.ml]: Just look at how beautiful
```OCaml
if daemon srv then
  `Daemonize_with (log_path srv)
else
  `Do_not_daemonize
```
could be

**Comment by smondet** [comment]
Merged into `master`.

File: src/lib/ketrew_configuration.ml [hunk 3]
```diff
@@ -165,98 +162,156 @@ let server
...
+            Buffer.add_string buf (line ^ "\n");
+            get_all ()
+          with e -> ()
+          end;
+        in
+        get_all ();
+        Buffer.contents buf in
+      close_in i;
+      content
+    with e -> close_in i; raise e
```
Comment (rleonid) [inline src/lib/ketrew_configuration.ml]: What exceptions can would escape the inner ones? Seems like you're guarding for `Buffer.contents buf` ?

Comment (smondet) [inline src/lib/ketrew_configuration.ml]: ah yes, I was moving around code and forgot to clean-up :D

Comment (smondet) [inline src/lib/ketrew_configuration.ml]: will be in the 3rd PR after this one :)
https://github.com/hammerlab/ketrew/tree/fix-useless-try-with

File: src/lib/ketrew_configuration.mli [hunk 4]
```diff
@@ -180,3 +142,47 @@ val connection: client -> string
...
+
+val load_exn:
+  ?and_apply:bool ->
+  ?profile:string ->
+  [ `From_path of string
+  | `Guess
+  | `In_directory of string
+  | `Override of t ] ->
+  t
+(** Load a configuration.
```
Comment (rleonid) [inline src/lib/ketrew_configuration.mli]: This documentation is confusing.
- I think describing the optional arguments later would be easier,.
- Why don't the order of the arguments match?
- `` `Override c `` seems super awkward as an argument to `load_exn`

File: src/lib/ketrew_plugin.ml [hunk 5]
```diff
@@ -49,31 +50,32 @@ let ketrew_deep_ancestors () =
   Findlib.package_deep_ancestors ["native"]
     (Lazy.force Ketrew_metadata.findlib_packages)
 
+let package_to_files_to_load package =
```
Comment (rleonid) [inline src/lib/ketrew_plugin.ml]: Name seems awkward, `package_files_to_load` ?

**Comment by rleonid** [comment]
Finished reviewing fwiw.

File: src/lib/ketrew_configuration.mli (hunk 4 above)
Comment (smondet) [inline src/lib/ketrew_configuration.mli]: ok for putting the optional arguments after, will do.

I don't understand the second point (?)

putting ``` `Override c``` outside of `load_exn` would mean more work for the callers of the function (every time have to figure out what to do). How would you organize that?

File: src/lib/ketrew_plugin.ml (hunk 5 above)
Comment (smondet) [inline src/lib/ketrew_plugin.ml]: `files_to_load_from_package` ?

File: src/lib/ketrew_configuration.mli (hunk 4 above)
Comment (rleonid) [inline src/lib/ketrew_configuration.mli]: 2nd point:
```OCaml
[ `From_path of string
| `Guess
| `In_directory of string
| `Override of t ] ->
```
vs
```OCaml
(*
- [`Override c]: use [c] as configuration
- [`From_path path]: parse the file [path]
- [`In_directory root]: look for configuration files in the [root] directory
- [`Guess]: use environment variables and/or default values to find the configuration file.
*)

```

Comment (rleonid) [inline src/lib/ketrew_configuration.mli]: 3rd point. If the callers have a configuration don't load it. Whatever __side effect__ is being performed by _loading_ should be exposed as a separate function, or refactored entirely.

File: src/lib/ketrew_plugin.ml (hunk 5 above)
Comment (rleonid) [inline src/lib/ketrew_plugin.ml]: fine.

File: src/lib/ketrew_configuration.mli (hunk 4 above)
Comment (smondet) [inline src/lib/ketrew_configuration.mli]: OK I see what order you meant.

The goal was to hide the side effects, in order to “one day” try to make them less global (which I don't know if it's even possible).

File: src/lib/ketrew_plugin.ml (hunk 5 above)
Comment (smondet) [inline src/lib/ketrew_plugin.ml]: Done: https://github.com/hammerlab/ketrew/tree/fix-plugin-fun-name

**Pull request closed without merging**